File: mycollab-scheduler/src/main/java/com/esofthead/mycollab/schedule/jobs/ProjectSendingRelayEmailNotificationJob.java
package com.esofthead.mycollab.schedule.jobs;

import java.util.List;

import org.quartz.JobExecutionContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.config.BeanDefinition;
import org.springframework.context.annotation.Scope;
import org.springframework.scheduling.quartz.QuartzJobBean;
import org.springframework.stereotype.Component;

import com.esofthead.mycollab.common.MonitorTypeConstants;
import com.esofthead.mycollab.common.service.RelayEmailNotificationService;
import com.esofthead.mycollab.core.MyCollabException;
import com.esofthead.mycollab.core.utils.BeanUtility;
import com.esofthead.mycollab.module.project.domain.ProjectRelayEmailNotification;
import com.esofthead.mycollab.module.project.service.ProjectService;
import com.esofthead.mycollab.schedule.email.SendingRelayEmailNotificationAction;
import com.esofthead.mycollab.spring.ApplicationContextUtil;

@Component
@Scope(value = BeanDefinition.SCOPE_PROTOTYPE)
public class ProjectSendingRelayEmailNotificationJob extends QuartzJobBean {
	private static Logger log = LoggerFactory
			.getLogger(ProjectSendingRelayEmailNotificationJob.class);

	@Override
	protected void executeInternal(JobExecutionContext context) {
		ProjectService projectService = (ProjectService) ApplicationContextUtil
				.getSpringBean(ProjectService.class);
		List<ProjectRelayEmailNotification> relayEmaiNotifications = projectService
				.findProjectRelayEmailNotifications();

		RelayEmailNotificationService relayNotificationService = ApplicationContextUtil
				.getSpringBean(RelayEmailNotificationService.class);
		log.debug("Get " + relayEmaiNotifications.size()
				+ " relay email notifications");
		SendingRelayEmailNotificationAction emailNotificationAction = null;

		for (ProjectRelayEmailNotification notification : relayEmaiNotifications) {
			try {
				if (notification.getEmailhandlerbean() != null
						&& notification.getType().startsWith("Project")) {
					emailNotificationAction = (SendingRelayEmailNotificationAction) ApplicationContextUtil
							.getSpringBean(Class.forName(notification
									.getEmailhandlerbean()));

					if (emailNotificationAction != null) {
						try {
							if (MonitorTypeConstants.CREATE_ACTION
									.equals(notification.getAction())) {
								emailNotificationAction
										.sendNotificationForCreateAction(notification);
							} else if (MonitorTypeConstants.UPDATE_ACTION
									.equals(notification.getAction())) {
								emailNotificationAction
										.sendNotificationForUpdateAction(notification);
							} else if (MonitorTypeConstants.ADD_COMMENT_ACTION
									.equals(notification.getAction())) {
								emailNotificationAction
										.sendNotificationForCommentAction(notification);
							}

							log.debug("Finish process notification {}",
									BeanUtility.printBeanObj(notification));
							relayNotificationService.removeWithSession(
									notification.getId(), "",
									notification.getSaccountid());

						} catch (Exception e) {
							log.error("Error when sending notification email",
									e);
						}
					}
				}

			} catch (ClassNotFoundException ex) {
				throw new MyCollabException("no class found toget spring bean "
						+ ex.getMessage());
			}

		}
	}
}

File: mycollab-scheduler/src/main/scala/com/mycollab/schedule/jobs/ProjectSendingRelayEmailNotificationJob.scala
package com.mycollab.schedule.jobs

import com.mycollab.module.project.service.ProjectService
import com.mycollab.common.MonitorTypeConstants
import com.mycollab.common.dao.RelayEmailNotificationMapper
import com.mycollab.spring.AppContextUtil
import org.quartz.{DisallowConcurrentExecution, JobExecutionContext}
import org.slf4j.{Logger, LoggerFactory}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.context.annotation.{Profile, Scope}
import org.springframework.stereotype.Component

/**
  * @author MyCollab Ltd.
  * @since 4.6.0
  */
@Component
@Profile(Array("production"))
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
@DisallowConcurrentExecution
class ProjectSendingRelayEmailNotificationJob extends GenericQuartzJobBean {
  private val LOG: Logger = LoggerFactory.getLogger(classOf[ProjectSendingRelayEmailNotificationJob])

  @Autowired
  private val projectService: ProjectService = null

  @Autowired
  private val relayNotificationMapper: RelayEmailNotificationMapper = null

  def executeJob(context: JobExecutionContext) {
    import scala.collection.JavaConverters._
    val relayEmailNotifications = projectService.findProjectRelayEmailNotifications.asScala.toList
    for (notification <- relayEmailNotifications) {
      try {
        val mailServiceCls = MailServiceMap.service(notification.getType)
        if (mailServiceCls != null) {
          val emailNotificationAction = AppContextUtil.getSpringBean(mailServiceCls)
          if (emailNotificationAction != null) {
            notification.getAction match {
              case MonitorTypeConstants.CREATE_ACTION => emailNotificationAction.sendNotificationForCreateAction(notification)
              case MonitorTypeConstants.UPDATE_ACTION => emailNotificationAction.sendNotificationForUpdateAction(notification)
              case MonitorTypeConstants.ADD_COMMENT_ACTION => emailNotificationAction.sendNotificationForCommentAction(notification)
            }
          } else {
            LOG.error("Can not find the spring class " + mailServiceCls)
          }
        }
      }
      catch {
        case ex: Exception => LOG.error("Error while sending scheduler command", ex)
      } finally {
        relayNotificationMapper.deleteByPrimaryKey(notification.getId)
      }
    }
  }
}
